src/app/workflow/dag/wdl-viewer/wdl-viewer.service.ts

```diff
@@ -0,0 +1,87 @@
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+import { Injectable } from '@angular/core';
+
+import * as JSZip from 'jszip';
+import * as pipeline from 'pipeline-builder';
+import { Observable, from, forkJoin } from 'rxjs';
+import { switchMap } from 'rxjs/internal/operators';
+import { GA4GHFilesQuery } from '../../../shared/ga4gh-files/ga4gh-files.query';
+import { ExtendedWorkflow } from '../../../shared/models/ExtendedWorkflow';
+import { ToolDescriptor, ToolFile, WorkflowsService, WorkflowVersion } from '../../../shared/swagger';
+
+
+/**
+ * Service for creating WDL workflow visualizations with EPAM Pipeline Builder library
+ */
+@Injectable()
+export class WdlViewerService {
+  private zip: JSZip = new JSZip();
+  constructor(private gA4GHFilesQuery: GA4GHFilesQuery, protected workflowsService: WorkflowsService) {
+  }
+
+  getFiles(descriptorType: ToolDescriptor.TypeEnum): Observable<Array<ToolFile>> {
+    return this.gA4GHFilesQuery.getToolFiles(descriptorType, [ToolFile.FileTypeEnum.PRIMARYDESCRIPTOR,
+      ToolFile.FileTypeEnum.SECONDARYDESCRIPTOR]);
+  }
+
+  /**
+   * Driver function for creating workflow visualizations
+   *
+   * @param files
+   * @param workflow
+   * @param version
+   */
+  create(files: Array<ToolFile>, workflow: ExtendedWorkflow, version: WorkflowVersion): Observable<any> {
+    if (files.length > 1) {
+      return this.createMultiple(workflow, version);
+    } else {
+      return this.createSingle(workflow, version);
+    }
+  }
+
+  /**
+   * Creates WDL workflow visualization for single-file workflows
+   *
+   * @param workflow
+   * @param version
+   */
+  createSingle(workflow: ExtendedWorkflow, version: WorkflowVersion): Observable<any> {
+    return this.workflowsService.wdl(workflow.id, version.name).pipe(switchMap(prim => {
+      return from(pipeline.parse(prim.content));
+    }));
+  }
+
+  /**
+   * Creates WDL workflow visualization for multi-file workflows (with import files)
+   *
+   * @param workflow
+   * @param version
+   */
+  createMultiple(workflow: ExtendedWorkflow, version: WorkflowVersion): Observable<any> {
+    return forkJoin(this.workflowsService.wdl(workflow.id, version.name), this.workflowsService.secondaryWdl(workflow.id, version.name))
+      .pipe(
+        switchMap(res => {
+          // Store each secondary file in a zip object
+          res[1].forEach(file => this.zip.file(file.path, file.content));
+
+          return from(this.zip.generateAsync({type: 'blob'})).pipe(switchMap(zip => {
+            return from(pipeline.parse(res[0].content, {zipFile: zip}));
```

Comment:
What happens if there is an error inside `parse()` method? Invalid WDL, cannot fetch HTTP imports because of CORS, etc.

Reply:
If there is an error, `parse()` throws a `Promise.rejection` with message. The message is displayed if that's the case.

Reply:
Where is it caught?

Since there are no types for this code where this could be gleaned from, I think it would be good to have a comment regarding this behavior.

Reply:
Sounds good. The error is caught by the Observable subscribing to this function, which in this case is inside the `wdl-viewer.component` class.